[BUG] [nl2sql] TableRelationNode无限循环bug

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current Behavior

使用 nl2sql/stream/search 接口时
TableRelationNode 节点若调用 businessKnowledgeRecallService.getFieldByDataSetId(dataSetId) 或 semanticModelRecallService.getFieldByDataSetId(String.valueOf(agentId)) 发生异常，会在该节点无限循环

### Expected Behavior

_No response_

### Steps To Reproduce

我目前在调试H2数据库来学习 nl2ql
调用 nl2sql/stream/search 接口时（nl2sql/search接口无此问题）
会由于h2的int或tinyint 无法转换为 boolean，导致TableRelationNode 节点报错，然后在此节点无限循环

### Environment

```markdown
Spring AI Alibaba version(s):1.0.0.4-SNAPSHOT
```

### Debug logs

2025-09-06T12:52:11.373+08:00 DEBUG 10428 --- [onPool-worker-3] o.s.jdbc.core.JdbcTemplate               : Executing prepared SQL query
2025-09-06T12:52:11.373+08:00 DEBUG 10428 --- [onPool-worker-3] o.s.jdbc.core.JdbcTemplate               : Executing prepared SQL statement [SELECT
	business_term,
	description,
       	synonyms,
	is_recall,
	data_set_id
FROM business_knowledge WHERE data_set_id = ? AND is_recall = 1
]
2025-09-06T12:52:11.373+08:00 DEBUG 10428 --- [onPool-worker-3] o.s.jdbc.datasource.DataSourceUtils      : Fetching JDBC Connection from DataSource
2025-09-06T12:52:11.375+08:00 ERROR 10428 --- [onPool-worker-3] c.alibaba.cloud.ai.graph.CompiledGraph   : org.springframework.dao.InvalidDataAccessApiUsageException: PreparedStatementCallback; SQL [SELECT
	business_term,
	description,
       	synonyms,
	is_recall,
	data_set_id
FROM business_knowledge WHERE data_set_id = ? AND is_recall = 1
]; Feature not supported: "converting to class boolean" [50100-232]

java.util.concurrent.ExecutionException: org.springframework.dao.InvalidDataAccessApiUsageException: PreparedStatementCallback; SQL [SELECT
	business_term,
	description,
       	synonyms,
	is_recall,
	data_set_id
FROM business_knowledge WHERE data_set_id = ? AND is_recall = 1
]; Feature not supported: "converting to class boolean" [50100-232]
	at java.base/java.util.concurrent.CompletableFuture.wrapInExecutionException(CompletableFuture.java:347) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:442) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2119) ~[na:na]
	at com.alibaba.cloud.ai.graph.CompiledGraph$AsyncNodeGenerator.next(CompiledGraph.java:1022) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.graph.async.AsyncGenerator$WithEmbed.next(AsyncGenerator.java:154) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.graph.async.AsyncGeneratorOperators.forEachAsync(AsyncGeneratorOperators.java:99) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.controller.Nl2sqlForGraphController.lambda$streamSearch$0(Nl2sqlForGraphController.java:185) ~[classes/:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run$$$capture(CompletableFuture.java:1850) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1842) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.doExec$$$capture(ForkJoinTask.java:507) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1394) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1970) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187) ~[na:na]
Caused by: org.springframework.dao.InvalidDataAccessApiUsageException: PreparedStatementCallback; SQL [SELECT
	business_term,
	description,
       	synonyms,
	is_recall,
	data_set_id
FROM business_knowledge WHERE data_set_id = ? AND is_recall = 1
]; Feature not supported: "converting to class boolean" [50100-232]
	at org.springframework.jdbc.support.SQLExceptionSubclassTranslator.doTranslate(SQLExceptionSubclassTranslator.java:106) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:107) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1556) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:677) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:723) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:754) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:767) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:820) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at com.alibaba.cloud.ai.service.business.BusinessKnowledgeRecallService.getFieldByDataSetId(BusinessKnowledgeRecallService.java:47) ~[classes/:na]
	at com.alibaba.cloud.ai.node.TableRelationNode.apply(TableRelationNode.java:108) ~[classes/:na]
	at com.alibaba.cloud.ai.graph.action.AsyncNodeAction.lambda$node_async$0(AsyncNodeAction.java:50) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.graph.action.AsyncNodeActionWithConfig.lambda$of$0(AsyncNodeActionWithConfig.java:56) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.graph.CompiledGraph$AsyncNodeGenerator.evaluateAction(CompiledGraph.java:836) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	... 12 common frames omitted
Caused by: org.h2.jdbc.JdbcSQLFeatureNotSupportedException: Feature not supported: "converting to class boolean" [50100-232]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:568) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:489) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.message.DbException.get(DbException.java:223) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.message.DbException.get(DbException.java:199) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.message.DbException.getUnsupportedException(DbException.java:287) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.ValueToObjectConverter.valueToOther(ValueToObjectConverter.java:389) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.ValueToObjectConverter.valueToObject(ValueToObjectConverter.java:343) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.jdbc.JdbcResultSet.getObject(JdbcResultSet.java:4216) ~[h2-2.3.232.jar:2.3.232]
	at com.alibaba.druid.pool.DruidPooledResultSet.getObject(DruidPooledResultSet.java:1765) ~[druid-1.2.22.jar:na]
	at com.alibaba.cloud.ai.service.business.BusinessKnowledgeRecallService.lambda$0(BusinessKnowledgeRecallService.java:51) ~[classes/:na]
	at org.springframework.jdbc.core.RowMapperResultSetExtractor.extractData(RowMapperResultSetExtractor.java:94) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.RowMapperResultSetExtractor.extractData(RowMapperResultSetExtractor.java:61) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate$1.doInPreparedStatement(JdbcTemplate.java:733) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:658) ~[spring-jdbc-6.2.6.jar:6.2.6]
	... 21 common frames omitted

2025-09-06T12:52:11.376+08:00  INFO 10428 --- [onPool-worker-3] c.a.cloud.ai.node.TableRelationNode      : Entering TableRelationNode node
2025-09-06T12:52:11.376+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent_knowledge.agent_id, vectorType=column, topK=10
2025-09-06T12:52:11.740+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent_knowledge.agent_id
2025-09-06T12:52:11.740+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent_datasource.agent_id, vectorType=column, topK=10
2025-09-06T12:52:11.894+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent_datasource.agent_id
2025-09-06T12:52:11.894+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=datasource.id, vectorType=column, topK=10
2025-09-06T12:52:12.056+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: datasource.id
2025-09-06T12:52:12.056+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent.id, vectorType=column, topK=10
2025-09-06T12:52:12.215+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent.id
2025-09-06T12:52:12.215+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=chat_session.agent_id, vectorType=column, topK=10
2025-09-06T12:52:12.358+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: chat_session.agent_id
2025-09-06T12:52:12.358+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=semantic_model.agent_id, vectorType=column, topK=10
2025-09-06T12:52:12.496+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: semantic_model.agent_id
2025-09-06T12:52:12.496+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=chat_session.id, vectorType=column, topK=10
2025-09-06T12:52:12.643+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: chat_session.id
2025-09-06T12:52:12.643+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=business_knowledge.agent_id, vectorType=column, topK=10
2025-09-06T12:52:12.795+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: business_knowledge.agent_id
2025-09-06T12:52:12.795+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent_preset_question.agent_id, vectorType=column, topK=10
2025-09-06T12:52:12.946+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent_preset_question.agent_id
2025-09-06T12:52:12.946+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=chat_message.session_id, vectorType=column, topK=10
2025-09-06T12:52:13.149+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: chat_message.session_id
2025-09-06T12:52:13.149+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent_datasource.datasource_id, vectorType=column, topK=10
2025-09-06T12:52:13.303+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent_datasource.datasource_id
2025-09-06T12:52:13.303+08:00  INFO 10428 --- [onPool-worker-3] c.a.cloud.ai.node.TableRelationNode      : [TableRelationNode] Executing regular schema selection
2025-09-06T12:52:13.303+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.ai.service.base.BaseNl2SqlService  : Fine selecting schema for query: 语义模型个数 with 0 evidences
2025-09-06T12:52:13.304+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.ai.service.base.BaseNl2SqlService  : Calling LLM for schema fine selection
2025-09-06T12:52:15.172+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.ai.service.base.BaseNl2SqlService  : Fine selection completed: 10 -> 1 tables, selected tables: [semantic_model]
2025-09-06T12:58:45.958+08:00 DEBUG 10428 --- [onPool-worker-3] o.s.jdbc.core.JdbcTemplate               : Executing prepared SQL query
2025-09-06T12:58:45.958+08:00 DEBUG 10428 --- [onPool-worker-3] o.s.jdbc.core.JdbcTemplate               : Executing prepared SQL statement [SELECT
	business_term,
	description,
       	synonyms,
	is_recall,
	data_set_id
FROM business_knowledge WHERE data_set_id = ? AND is_recall = 1
]
2025-09-06T12:58:45.958+08:00 DEBUG 10428 --- [onPool-worker-3] o.s.jdbc.datasource.DataSourceUtils      : Fetching JDBC Connection from DataSource
2025-09-06T12:58:45.959+08:00 ERROR 10428 --- [onPool-worker-3] c.alibaba.cloud.ai.graph.CompiledGraph   : org.springframework.dao.InvalidDataAccessApiUsageException: PreparedStatementCallback; SQL [SELECT
	business_term,
	description,
       	synonyms,
	is_recall,
	data_set_id
FROM business_knowledge WHERE data_set_id = ? AND is_recall = 1
]; Feature not supported: "converting to class boolean" [50100-232]

java.util.concurrent.ExecutionException: org.springframework.dao.InvalidDataAccessApiUsageException: PreparedStatementCallback; SQL [SELECT
	business_term,
	description,
       	synonyms,
	is_recall,
	data_set_id
FROM business_knowledge WHERE data_set_id = ? AND is_recall = 1
]; Feature not supported: "converting to class boolean" [50100-232]
	at java.base/java.util.concurrent.CompletableFuture.wrapInExecutionException(CompletableFuture.java:347) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:442) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2119) ~[na:na]
	at com.alibaba.cloud.ai.graph.CompiledGraph$AsyncNodeGenerator.next(CompiledGraph.java:1022) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.graph.async.AsyncGenerator$WithEmbed.next(AsyncGenerator.java:154) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.graph.async.AsyncGeneratorOperators.forEachAsync(AsyncGeneratorOperators.java:99) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.controller.Nl2sqlForGraphController.lambda$streamSearch$0(Nl2sqlForGraphController.java:185) ~[classes/:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run$$$capture(CompletableFuture.java:1850) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1842) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.doExec$$$capture(ForkJoinTask.java:507) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1394) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1970) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187) ~[na:na]
Caused by: org.springframework.dao.InvalidDataAccessApiUsageException: PreparedStatementCallback; SQL [SELECT
	business_term,
	description,
       	synonyms,
	is_recall,
	data_set_id
FROM business_knowledge WHERE data_set_id = ? AND is_recall = 1
]; Feature not supported: "converting to class boolean" [50100-232]
	at org.springframework.jdbc.support.SQLExceptionSubclassTranslator.doTranslate(SQLExceptionSubclassTranslator.java:106) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:107) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1556) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:677) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:723) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:754) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:767) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:820) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at com.alibaba.cloud.ai.service.business.BusinessKnowledgeRecallService.getFieldByDataSetId(BusinessKnowledgeRecallService.java:47) ~[classes/:na]
	at com.alibaba.cloud.ai.node.TableRelationNode.apply(TableRelationNode.java:108) ~[classes/:na]
	at com.alibaba.cloud.ai.graph.action.AsyncNodeAction.lambda$node_async$0(AsyncNodeAction.java:50) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.graph.action.AsyncNodeActionWithConfig.lambda$of$0(AsyncNodeActionWithConfig.java:56) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	at com.alibaba.cloud.ai.graph.CompiledGraph$AsyncNodeGenerator.evaluateAction(CompiledGraph.java:836) ~[spring-ai-alibaba-graph-core-1.0.0.4-SNAPSHOT.jar:1.0.0.4-SNAPSHOT]
	... 12 common frames omitted
Caused by: org.h2.jdbc.JdbcSQLFeatureNotSupportedException: Feature not supported: "converting to class boolean" [50100-232]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:568) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:489) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.message.DbException.get(DbException.java:223) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.message.DbException.get(DbException.java:199) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.message.DbException.getUnsupportedException(DbException.java:287) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.ValueToObjectConverter.valueToOther(ValueToObjectConverter.java:389) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.value.ValueToObjectConverter.valueToObject(ValueToObjectConverter.java:343) ~[h2-2.3.232.jar:2.3.232]
	at org.h2.jdbc.JdbcResultSet.getObject(JdbcResultSet.java:4216) ~[h2-2.3.232.jar:2.3.232]
	at com.alibaba.druid.pool.DruidPooledResultSet.getObject(DruidPooledResultSet.java:1765) ~[druid-1.2.22.jar:na]
	at com.alibaba.cloud.ai.service.business.BusinessKnowledgeRecallService.lambda$0(BusinessKnowledgeRecallService.java:51) ~[classes/:na]
	at org.springframework.jdbc.core.RowMapperResultSetExtractor.extractData(RowMapperResultSetExtractor.java:94) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.RowMapperResultSetExtractor.extractData(RowMapperResultSetExtractor.java:61) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate$1.doInPreparedStatement(JdbcTemplate.java:733) ~[spring-jdbc-6.2.6.jar:6.2.6]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:658) ~[spring-jdbc-6.2.6.jar:6.2.6]
	... 21 common frames omitted

2025-09-06T12:58:45.959+08:00  INFO 10428 --- [onPool-worker-3] c.a.cloud.ai.node.TableRelationNode      : Entering TableRelationNode node
2025-09-06T12:58:45.960+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent_knowledge.agent_id, vectorType=column, topK=10
2025-09-06T12:58:46.233+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent_knowledge.agent_id
2025-09-06T12:58:46.233+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent_datasource.agent_id, vectorType=column, topK=10
2025-09-06T12:58:46.381+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent_datasource.agent_id
2025-09-06T12:58:46.381+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=datasource.id, vectorType=column, topK=10
2025-09-06T12:58:46.518+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: datasource.id
2025-09-06T12:58:46.518+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent.id, vectorType=column, topK=10
2025-09-06T12:58:46.688+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent.id
2025-09-06T12:58:46.688+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=chat_session.agent_id, vectorType=column, topK=10
2025-09-06T12:58:46.822+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: chat_session.agent_id
2025-09-06T12:58:46.823+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=semantic_model.agent_id, vectorType=column, topK=10
2025-09-06T12:58:46.998+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: semantic_model.agent_id
2025-09-06T12:58:46.998+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=chat_session.id, vectorType=column, topK=10
2025-09-06T12:58:47.148+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: chat_session.id
2025-09-06T12:58:47.148+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=business_knowledge.agent_id, vectorType=column, topK=10
2025-09-06T12:58:47.296+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: business_knowledge.agent_id
2025-09-06T12:58:47.296+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent_preset_question.agent_id, vectorType=column, topK=10
2025-09-06T12:58:47.456+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent_preset_question.agent_id
2025-09-06T12:58:47.457+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=chat_message.session_id, vectorType=column, topK=10
2025-09-06T12:58:47.581+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: chat_message.session_id
2025-09-06T12:58:47.581+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Searching table by name and vectorType: name=agent_datasource.datasource_id, vectorType=column, topK=10
2025-09-06T12:58:47.738+08:00  INFO 10428 --- [onPool-worker-3] c.a.c.a.s.s.SimpleVectorStoreService     : Search by name completed. Found 0 documents for name: agent_datasource.datasource_id
2025-09-06T12:58:47.738+08:00  INFO 10428 --- [onPool-worker-3] c.a.cloud.ai.node.TableRelationNode      : [TableRelationNode] Executing regular schema selection
2025-09-06T12:58:47.738+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.ai.service.base.BaseNl2SqlService  : Fine selecting schema for query: 语义模型个数 with 0 evidences
2025-09-06T12:58:47.739+08:00 DEBUG 10428 --- [onPool-worker-3] c.a.c.ai.service.base.BaseNl2SqlService  : Calling LLM for schema fine selection


### Anything else?

_No response_

--- a/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/main/java/com/alibaba/cloud/ai/config/Nl2sqlConfiguration.java
+++ b/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/main/java/com/alibaba/cloud/ai/config/Nl2sqlConfiguration.java
@@ -19,38 +19,21 @@ package com.alibaba.cloud.ai.config;
 import com.alibaba.cloud.ai.connector.accessor.Accessor;
 import com.alibaba.cloud.ai.connector.config.DbConfig;
 import com.alibaba.cloud.ai.constant.Constant;
-import com.alibaba.cloud.ai.dispatcher.PlanExecutorDispatcher;
-import com.alibaba.cloud.ai.dispatcher.PythonExecutorDispatcher;
-import com.alibaba.cloud.ai.dispatcher.QueryRewriteDispatcher;
-import com.alibaba.cloud.ai.dispatcher.SQLExecutorDispatcher;
-import com.alibaba.cloud.ai.dispatcher.SemanticConsistenceDispatcher;
-import com.alibaba.cloud.ai.dispatcher.SqlGenerateDispatcher;
+import com.alibaba.cloud.ai.dispatcher.*;
 import com.alibaba.cloud.ai.graph.GraphRepresentation;
 import com.alibaba.cloud.ai.graph.KeyStrategy;
 import com.alibaba.cloud.ai.graph.KeyStrategyFactory;
 import com.alibaba.cloud.ai.graph.StateGraph;
 import com.alibaba.cloud.ai.graph.exception.GraphStateException;
 import com.alibaba.cloud.ai.graph.state.strategy.ReplaceStrategy;
-import com.alibaba.cloud.ai.node.KeywordExtractNode;
-import com.alibaba.cloud.ai.node.PlanExecutorNode;
-import com.alibaba.cloud.ai.node.PlannerNode;
-import com.alibaba.cloud.ai.node.PythonAnalyzeNode;
-import com.alibaba.cloud.ai.node.PythonExecuteNode;
-import com.alibaba.cloud.ai.node.PythonGenerateNode;
-import com.alibaba.cloud.ai.node.QueryRewriteNode;
-import com.alibaba.cloud.ai.node.ReportGeneratorNode;
-import com.alibaba.cloud.ai.node.SchemaRecallNode;
-import com.alibaba.cloud.ai.node.SemanticConsistencyNode;
-import com.alibaba.cloud.ai.node.SqlExecuteNode;
-import com.alibaba.cloud.ai.node.SqlGenerateNode;
-import com.alibaba.cloud.ai.node.TableRelationNode;
+import com.alibaba.cloud.ai.node.*;
 import com.alibaba.cloud.ai.service.DatasourceService;
+import com.alibaba.cloud.ai.service.UserPromptConfigService;
 import com.alibaba.cloud.ai.service.base.BaseNl2SqlService;
 import com.alibaba.cloud.ai.service.base.BaseSchemaService;
 import com.alibaba.cloud.ai.service.business.BusinessKnowledgeRecallService;
-import com.alibaba.cloud.ai.service.semantic.SemanticModelRecallService;
 import com.alibaba.cloud.ai.service.code.CodePoolExecutorService;
-import com.alibaba.cloud.ai.service.UserPromptConfigService;
+import com.alibaba.cloud.ai.service.semantic.SemanticModelRecallService;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.ai.chat.client.ChatClient;
@@ -61,54 +44,7 @@ import org.springframework.context.annotation.Configuration;
 import java.util.HashMap;
 import java.util.Map;
 
-import static com.alibaba.cloud.ai.constant.Constant.AGENT_ID;
-import static com.alibaba.cloud.ai.constant.Constant.BUSINESS_KNOWLEDGE;
-import static com.alibaba.cloud.ai.constant.Constant.COLUMN_DOCUMENTS_BY_KEYWORDS_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.EVIDENCES;
-import static com.alibaba.cloud.ai.constant.Constant.INPUT_KEY;
-import static com.alibaba.cloud.ai.constant.Constant.IS_ONLY_NL2SQL;
-import static com.alibaba.cloud.ai.constant.Constant.KEYWORD_EXTRACT_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.KEYWORD_EXTRACT_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.NL2SQL_GRAPH_NAME;
-import static com.alibaba.cloud.ai.constant.Constant.ONLY_NL2SQL_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.PLANNER_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.PLANNER_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.PLAN_CURRENT_STEP;
-import static com.alibaba.cloud.ai.constant.Constant.PLAN_EXECUTOR_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.PLAN_NEXT_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.PLAN_REPAIR_COUNT;
-import static com.alibaba.cloud.ai.constant.Constant.PLAN_VALIDATION_ERROR;
-import static com.alibaba.cloud.ai.constant.Constant.PLAN_VALIDATION_STATUS;
-import static com.alibaba.cloud.ai.constant.Constant.PYTHON_ANALYSIS_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.PYTHON_ANALYZE_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.PYTHON_EXECUTE_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.PYTHON_EXECUTE_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.PYTHON_GENERATE_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.PYTHON_GENERATE_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.PYTHON_IS_SUCCESS;
-import static com.alibaba.cloud.ai.constant.Constant.PYTHON_TRIES_COUNT;
-import static com.alibaba.cloud.ai.constant.Constant.QUERY_REWRITE_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.QUERY_REWRITE_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.REPORT_GENERATOR_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.RESULT;
-import static com.alibaba.cloud.ai.constant.Constant.SCHEMA_RECALL_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.SEMANTIC_CONSISTENCY_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.SEMANTIC_CONSISTENCY_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.SEMANTIC_CONSISTENCY_NODE_RECOMMEND_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.SEMANTIC_MODEL;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_EXECUTE_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_EXECUTE_NODE_EXCEPTION_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_EXECUTE_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_GENERATE_COUNT;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_GENERATE_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_GENERATE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_GENERATE_SCHEMA_MISSING_ADVICE;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_RESULT_LIST_MEMORY;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_VALIDATE_EXCEPTION_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_VALIDATE_NODE_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.TABLE_DOCUMENTS_FOR_SCHEMA_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.TABLE_RELATION_NODE;
-import static com.alibaba.cloud.ai.constant.Constant.TABLE_RELATION_OUTPUT;
+import static com.alibaba.cloud.ai.constant.Constant.*;
 import static com.alibaba.cloud.ai.graph.StateGraph.END;
 import static com.alibaba.cloud.ai.graph.StateGraph.START;
 import static com.alibaba.cloud.ai.graph.action.AsyncEdgeAction.edge_async;
@@ -189,6 +125,8 @@ public class Nl2sqlConfiguration {
 			keyStrategyHashMap.put(SQL_VALIDATE_EXCEPTION_OUTPUT, new ReplaceStrategy());
 			// table relation节点输出
 			keyStrategyHashMap.put(TABLE_RELATION_OUTPUT, new ReplaceStrategy());
+			keyStrategyHashMap.put(TABLE_RELATION_EXCEPTION_OUTPUT, new ReplaceStrategy());
+			keyStrategyHashMap.put(TABLE_RELATION_RETRY_COUNT, new ReplaceStrategy());
 			// sql generate节点输出
 			keyStrategyHashMap.put(SQL_GENERATE_SCHEMA_MISSING_ADVICE, new ReplaceStrategy());
 			keyStrategyHashMap.put(SQL_GENERATE_OUTPUT, new ReplaceStrategy());
@@ -245,7 +183,10 @@ public class Nl2sqlConfiguration {
 					Map.of(KEYWORD_EXTRACT_NODE, KEYWORD_EXTRACT_NODE, END, END))
 			.addEdge(KEYWORD_EXTRACT_NODE, SCHEMA_RECALL_NODE)
 			.addEdge(SCHEMA_RECALL_NODE, TABLE_RELATION_NODE)
-			.addEdge(TABLE_RELATION_NODE, PLANNER_NODE)
+			.addConditionalEdges(TABLE_RELATION_NODE, edge_async(new TableRelationDispatcher()),
+					Map.of(PLANNER_NODE, PLANNER_NODE, END, END, TABLE_RELATION_NODE, TABLE_RELATION_NODE))// retry
+																											// table
+																											// relation
 			// The edge from PlannerNode now goes to PlanExecutorNode for validation and
 			// execution
 			.addEdge(PLANNER_NODE, PLAN_EXECUTOR_NODE)

--- a/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/main/java/com/alibaba/cloud/ai/constant/Constant.java
+++ b/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/main/java/com/alibaba/cloud/ai/constant/Constant.java
@@ -43,6 +43,10 @@ public class Constant {
 
 	public static final String TABLE_RELATION_OUTPUT = "TABLE_RELATION_OUTPUT";
 
+	public static final String TABLE_RELATION_EXCEPTION_OUTPUT = "TABLE_RELATION_EXCEPTION_OUTPUT";
+
+	public static final String TABLE_RELATION_RETRY_COUNT = "TABLE_RELATION_RETRY_COUNT";
+
 	public static final String BUSINESS_KNOWLEDGE = "BUSINESS_KNOWLEDGE";
 
 	public static final String SEMANTIC_MODEL = "SEMANTIC_MODEL";

--- a/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/main/java/com/alibaba/cloud/ai/dispatcher/TableRelationDispatcher.java
+++ b/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/main/java/com/alibaba/cloud/ai/dispatcher/TableRelationDispatcher.java
@@ -1,0 +1,60 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.cloud.ai.dispatcher;
+
+import com.alibaba.cloud.ai.graph.OverAllState;
+import com.alibaba.cloud.ai.graph.action.EdgeAction;
+import com.alibaba.cloud.ai.util.StateUtils;
+
+import java.util.Optional;
+
+import static com.alibaba.cloud.ai.constant.Constant.*;
+import static com.alibaba.cloud.ai.graph.StateGraph.END;
+
+public class TableRelationDispatcher implements EdgeAction {
+
+	private static final int MAX_RETRY_COUNT = 3;
+
+	@Override
+	public String apply(OverAllState state) throws Exception {
+
+		String errorFlag = StateUtils.getStringValue(state, TABLE_RELATION_EXCEPTION_OUTPUT, null);
+		Integer retryCount = StateUtils.getObjectValue(state, TABLE_RELATION_RETRY_COUNT, Integer.class, 0);
+
+		if (errorFlag != null && !errorFlag.isEmpty()) {
+			if (isRetryableError(errorFlag) && retryCount < MAX_RETRY_COUNT) {
+				return TABLE_RELATION_NODE;
+			}
+			else {
+				return END;
+			}
+		}
+
+		Optional<String> tableRelationOutput = state.value(TABLE_RELATION_OUTPUT);
+		if (tableRelationOutput.isPresent()) {
+			return PLANNER_NODE; // next node is planner
+		}
+
+		// no output, end
+		return END;
+	}
+
+	private boolean isRetryableError(String errorMessage) {
+		return errorMessage.startsWith("RETRYABLE:");
+	}
+
+}

--- a/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/main/java/com/alibaba/cloud/ai/node/TableRelationNode.java
+++ b/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/main/java/com/alibaba/cloud/ai/node/TableRelationNode.java
@@ -19,10 +19,10 @@ package com.alibaba.cloud.ai.node;
 import com.alibaba.cloud.ai.constant.Constant;
 import com.alibaba.cloud.ai.dto.BusinessKnowledgeDTO;
 import com.alibaba.cloud.ai.dto.SemanticModelDTO;
+import com.alibaba.cloud.ai.dto.schema.SchemaDTO;
 import com.alibaba.cloud.ai.enums.StreamResponseType;
 import com.alibaba.cloud.ai.graph.OverAllState;
 import com.alibaba.cloud.ai.graph.action.NodeAction;
-import com.alibaba.cloud.ai.dto.schema.SchemaDTO;
 import com.alibaba.cloud.ai.service.base.BaseNl2SqlService;
 import com.alibaba.cloud.ai.service.base.BaseSchemaService;
 import com.alibaba.cloud.ai.service.business.BusinessKnowledgeRecallService;
@@ -34,21 +34,13 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.ai.chat.model.ChatResponse;
 import org.springframework.ai.document.Document;
+import org.springframework.dao.DataAccessException;
 import reactor.core.publisher.Flux;
 
 import java.util.List;
 import java.util.Map;
 
-import static com.alibaba.cloud.ai.constant.Constant.AGENT_ID;
-import static com.alibaba.cloud.ai.constant.Constant.BUSINESS_KNOWLEDGE;
-import static com.alibaba.cloud.ai.constant.Constant.COLUMN_DOCUMENTS_BY_KEYWORDS_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.EVIDENCES;
-import static com.alibaba.cloud.ai.constant.Constant.INPUT_KEY;
-import static com.alibaba.cloud.ai.constant.Constant.SEMANTIC_MODEL;
-import static com.alibaba.cloud.ai.constant.Constant.SQL_GENERATE_SCHEMA_MISSING_ADVICE;
-import static com.alibaba.cloud.ai.constant.Constant.TABLE_DOCUMENTS_FOR_SCHEMA_OUTPUT;
-import static com.alibaba.cloud.ai.constant.Constant.TABLE_RELATION_OUTPUT;
-
+import static com.alibaba.cloud.ai.constant.Constant.*;
 import static com.alibaba.cloud.ai.prompt.PromptHelper.buildBusinessKnowledgePrompt;
 import static com.alibaba.cloud.ai.prompt.PromptHelper.buildSemanticModelPrompt;
 
@@ -87,6 +79,8 @@ public class TableRelationNode implements NodeAction {
 	public Map<String, Object> apply(OverAllState state) throws Exception {
 		logger.info("Entering {} node", this.getClass().getSimpleName());
 
+		int retryCount = StateUtils.getObjectValue(state, TABLE_RELATION_RETRY_COUNT, Integer.class, 0);
+
 		// Get necessary input parameters
 		String input = StateUtils.getStringValue(state, INPUT_KEY);
 		List<String> evidenceList = StateUtils.getListValue(state, EVIDENCES);
@@ -104,9 +98,20 @@ public class TableRelationNode implements NodeAction {
 		SchemaDTO schemaDTO = buildInitialSchema(columnDocumentsByKeywords, tableDocuments);
 		SchemaDTO result = processSchemaSelection(schemaDTO, input, evidenceList, state);
 
-		// Extract business knowledge and semantic model
-		List<BusinessKnowledgeDTO> businessKnowledges = businessKnowledgeRecallService.getFieldByDataSetId(dataSetId);
-		List<SemanticModelDTO> semanticModel = semanticModelRecallService.getFieldByDataSetId(String.valueOf(agentId));
+		List<BusinessKnowledgeDTO> businessKnowledges;
+		List<SemanticModelDTO> semanticModel;
+		try {
+			// Extract business knowledge and semantic model
+			businessKnowledges = businessKnowledgeRecallService.getFieldByDataSetId(dataSetId);
+			semanticModel = semanticModelRecallService.getFieldByDataSetId(String.valueOf(agentId));
+		}
+		catch (DataAccessException e) {
+			logger.warn("Database query failed (attempt {}): {}", retryCount + 1, e.getMessage());
+
+			String errorType = classifyDatabaseError(e);
+			return Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, errorType + ": " + e.getMessage(),
+					TABLE_RELATION_RETRY_COUNT, retryCount + 1);
+		}
 		// load prompt template
 		String businessKnowledgePrompt = buildBusinessKnowledgePrompt(businessKnowledges);
 		String semanticModelPrompt = buildSemanticModelPrompt(semanticModel);
@@ -130,8 +135,21 @@ public class TableRelationNode implements NodeAction {
 						businessKnowledgePrompt, SEMANTIC_MODEL, semanticModelPrompt),
 				displayFlux, StreamResponseType.SCHEMA_DEEP_RECALL);
 
+		// need to reset retry count and exception
 		return Map.of(TABLE_RELATION_OUTPUT, generator, BUSINESS_KNOWLEDGE, businessKnowledgePrompt, SEMANTIC_MODEL,
-				semanticModelPrompt);
+				semanticModelPrompt, TABLE_RELATION_RETRY_COUNT, 0, TABLE_RELATION_EXCEPTION_OUTPUT, "");
+
+	}
+
+	private String classifyDatabaseError(DataAccessException e) {
+		String message = e.getMessage();
+		if (message != null) {
+			// timeout, connection, network can be retried
+			if (message.contains("timeout") || message.contains("connection") || message.contains("network")) {
+				return "RETRYABLE";
+			}
+		}
+		return "NON_RETRYABLE";
 	}
 
 	/**

--- a/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/test/java/com/alibaba/cloud/ai/dispatcher/TableRelationDispatcherTest.java
+++ b/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/test/java/com/alibaba/cloud/ai/dispatcher/TableRelationDispatcherTest.java
@@ -1,0 +1,187 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.cloud.ai.dispatcher;
+
+import com.alibaba.cloud.ai.graph.OverAllState;
+import com.alibaba.cloud.ai.graph.state.strategy.ReplaceStrategy;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.Map;
+
+import static com.alibaba.cloud.ai.constant.Constant.*;
+import static com.alibaba.cloud.ai.graph.StateGraph.END;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class TableRelationDispatcherTest {
+
+	private TableRelationDispatcher dispatcher;
+
+	private OverAllState state;
+
+	@BeforeEach
+	void setUp() {
+		dispatcher = new TableRelationDispatcher();
+
+		// Initialize state
+		state = new OverAllState();
+		state.registerKeyAndStrategy(TABLE_RELATION_EXCEPTION_OUTPUT, new ReplaceStrategy());
+		state.registerKeyAndStrategy(TABLE_RELATION_RETRY_COUNT, new ReplaceStrategy());
+		state.registerKeyAndStrategy(TABLE_RELATION_OUTPUT, new ReplaceStrategy());
+	}
+
+	@Test
+	void testSuccessfulExecution() throws Exception {
+		// Set success state
+		state.updateState(Map.of(TABLE_RELATION_OUTPUT, "mock_generator_output"));
+
+		// Execute test
+		String result = dispatcher.apply(state);
+
+		// Verify routing decision
+		assertEquals(PLANNER_NODE, result);
+	}
+
+	@Test
+	void testRetryableError_WithinRetryLimit() throws Exception {
+		// Set retryable error, retry count not reached limit
+		state.updateState(Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, "RETRYABLE: Connection timeout",
+				TABLE_RELATION_RETRY_COUNT, 2));
+
+		// execute test
+		String result = dispatcher.apply(state);
+
+		// Verify should retry
+		assertEquals(TABLE_RELATION_NODE, result);
+	}
+
+	@Test
+	void testRetryableError_ExceedsRetryLimit() throws Exception {
+		// Set retryable error, but retry count reached limit
+		state.updateState(Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, "RETRYABLE: Connection timeout",
+				TABLE_RELATION_RETRY_COUNT, 3));
+
+		// execute test
+		String result = dispatcher.apply(state);
+
+		// Verify should terminate
+		assertEquals(END, result);
+	}
+
+	@Test
+	void testNonRetryableError() throws Exception {
+		// Set non-retryable error
+		state.updateState(Map.of(TABLE_RELATION_EXCEPTION_OUTPUT,
+				"NON_RETRYABLE: Feature not supported: converting to class boolean", TABLE_RELATION_RETRY_COUNT, 1));
+
+		// execute test
+		String result = dispatcher.apply(state);
+
+		// Verify should terminate
+		assertEquals(END, result);
+	}
+
+	@Test
+	void testNoOutput_NoError() throws Exception {
+		// Set no output no error state
+		state.updateState(Map.of());
+
+		// execute test
+		String result = dispatcher.apply(state);
+
+		// Verify should terminate
+		assertEquals(END, result);
+	}
+
+	@Test
+	void testRetryCountBoundary() throws Exception {
+		// Test retry count boundary case
+		state.updateState(
+				Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, "RETRYABLE: Network error", TABLE_RELATION_RETRY_COUNT, 2 // Just
+																													// within
+																													// the
+																													// limit
+				));
+
+		String result = dispatcher.apply(state);
+		assertEquals(TABLE_RELATION_NODE, result);
+
+		// Test reaching the limit
+		state.updateState(
+				Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, "RETRYABLE: Network error", TABLE_RELATION_RETRY_COUNT, 3 // 达到限制
+				));
+
+		result = dispatcher.apply(state);
+		assertEquals(END, result);
+	}
+
+	@Test
+	void testMixedErrorTypes() throws Exception {
+		// Test handling of different error types
+		String[] retryableErrors = { "RETRYABLE: timeout", "RETRYABLE: connection" };
+
+		String[] nonRetryableErrors = { "NON_RETRYABLE: boolean conversion", "NON_RETRYABLE: configuration error" };
+
+		// Test retryable errors
+		for (String errorType : retryableErrors) {
+			state.updateState(Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, errorType, TABLE_RELATION_RETRY_COUNT, 1));
+
+			String result = dispatcher.apply(state);
+			assertEquals(TABLE_RELATION_NODE, result, "Should retry for: " + errorType);
+		}
+
+		// Test non-retryable errors
+		for (String errorType : nonRetryableErrors) {
+			state.updateState(Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, errorType, TABLE_RELATION_RETRY_COUNT, 1));
+
+			String result = dispatcher.apply(state);
+			assertEquals(END, result, "Should end for: " + errorType);
+		}
+	}
+
+	@Test
+	void testZeroRetryCount() throws Exception {
+		// Test case with retry count 0
+		state.updateState(Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, "RETRYABLE: Connection timeout",
+				TABLE_RELATION_RETRY_COUNT, 0));
+
+		String result = dispatcher.apply(state);
+		assertEquals(TABLE_RELATION_NODE, result);
+	}
+
+	@Test
+	void testNullRetryCount() throws Exception {
+		// Test case with null retry count (should be treated as 0)
+		state.updateState(Map.of(TABLE_RELATION_EXCEPTION_OUTPUT, "RETRYABLE: Connection timeout"
+		// 不设置 TABLE_RELATION_RETRY_COUNT
+		));
+
+		String result = dispatcher.apply(state);
+		assertEquals(TABLE_RELATION_NODE, result);
+	}
+
+	@Test
+	void testSuccessWithPreviousError() throws Exception {
+		// Test case with previous error but now successful
+		state.updateState(Map.of(TABLE_RELATION_OUTPUT, "success_output", TABLE_RELATION_EXCEPTION_OUTPUT, "", // 错误已清除
+				TABLE_RELATION_RETRY_COUNT, 0));
+
+		String result = dispatcher.apply(state);
+		assertEquals(PLANNER_NODE, result);
+	}
+
+}

--- a/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/test/java/com/alibaba/cloud/ai/node/TableRelationNodeTest.java
+++ b/spring-ai-alibaba-nl2sql/spring-ai-alibaba-nl2sql-chat/src/test/java/com/alibaba/cloud/ai/node/TableRelationNodeTest.java
@@ -1,0 +1,243 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.cloud.ai.node;
+
+import com.alibaba.cloud.ai.dto.BusinessKnowledgeDTO;
+import com.alibaba.cloud.ai.dto.SemanticModelDTO;
+import com.alibaba.cloud.ai.dto.schema.SchemaDTO;
+import com.alibaba.cloud.ai.graph.OverAllState;
+import com.alibaba.cloud.ai.graph.state.strategy.ReplaceStrategy;
+import com.alibaba.cloud.ai.service.base.BaseNl2SqlService;
+import com.alibaba.cloud.ai.service.base.BaseSchemaService;
+import com.alibaba.cloud.ai.service.business.BusinessKnowledgeRecallService;
+import com.alibaba.cloud.ai.service.semantic.SemanticModelRecallService;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.ai.document.Document;
+import org.springframework.dao.DataAccessException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static com.alibaba.cloud.ai.constant.Constant.*;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class TableRelationNodeTest {
+
+	@Mock
+	private BaseSchemaService baseSchemaService;
+
+	@Mock
+	private BaseNl2SqlService baseNl2SqlService;
+
+	@Mock
+	private BusinessKnowledgeRecallService businessKnowledgeRecallService;
+
+	@Mock
+	private SemanticModelRecallService semanticModelRecallService;
+
+	private TableRelationNode tableRelationNode;
+
+	private OverAllState state;
+
+	@BeforeEach
+	void setUp() {
+		tableRelationNode = new TableRelationNode(baseSchemaService, baseNl2SqlService, businessKnowledgeRecallService,
+				semanticModelRecallService);
+
+		// Initialize state
+		state = new OverAllState();
+		state.registerKeyAndStrategy(INPUT_KEY, new ReplaceStrategy());
+		state.registerKeyAndStrategy(EVIDENCES, new ReplaceStrategy());
+		state.registerKeyAndStrategy(TABLE_DOCUMENTS_FOR_SCHEMA_OUTPUT, new ReplaceStrategy());
+		state.registerKeyAndStrategy(COLUMN_DOCUMENTS_BY_KEYWORDS_OUTPUT, new ReplaceStrategy());
+		state.registerKeyAndStrategy(AGENT_ID, new ReplaceStrategy());
+		state.registerKeyAndStrategy(TABLE_RELATION_RETRY_COUNT, new ReplaceStrategy());
+		state.registerKeyAndStrategy(TABLE_RELATION_EXCEPTION_OUTPUT, new ReplaceStrategy());
+		state.registerKeyAndStrategy(TABLE_RELATION_OUTPUT, new ReplaceStrategy());
+		state.registerKeyAndStrategy(BUSINESS_KNOWLEDGE, new ReplaceStrategy());
+		state.registerKeyAndStrategy(SEMANTIC_MODEL, new ReplaceStrategy());
+	}
+
+	@Test
+	void testSuccessfulExecution() throws Exception {
+		// Prepare test data
+		setupSuccessfulMocks();
+		setupStateWithValidData();
+
+		// Execute test
+		Map<String, Object> result = tableRelationNode.apply(state);
+
+		// Verify results
+		assertNotNull(result);
+		assertTrue(result.containsKey(TABLE_RELATION_OUTPUT));
+		assertTrue(result.containsKey(BUSINESS_KNOWLEDGE));
+		assertTrue(result.containsKey(SEMANTIC_MODEL));
+		assertEquals(0, result.get(TABLE_RELATION_RETRY_COUNT));
+		assertEquals("", result.get(TABLE_RELATION_EXCEPTION_OUTPUT));
+
+		// Verify service calls, check TableRelationNode code uses agentId value for
+		// DataSetId
+		verify(businessKnowledgeRecallService).getFieldByDataSetId("123");
+		verify(semanticModelRecallService).getFieldByDataSetId("123");
+		verify(baseSchemaService).extractDatabaseName(any(SchemaDTO.class));
+		verify(baseNl2SqlService).fineSelect(any(SchemaDTO.class), anyString(), any(List.class));
+	}
+
+	@Test
+	void testDatabaseQueryException_Retryable() throws Exception {
+		// Simulate retryable database exception
+		DataAccessException retryableException = new DataAccessException("Connection timeout") {
+		};
+		when(businessKnowledgeRecallService.getFieldByDataSetId(anyString())).thenThrow(retryableException);
+
+		setupStateWithValidData();
+		setupSchemaServiceMocks();
+
+		// execute test
+		Map<String, Object> result = tableRelationNode.apply(state);
+
+		// Verify error handling
+		assertNotNull(result);
+		assertTrue(result.containsKey(TABLE_RELATION_EXCEPTION_OUTPUT));
+		assertTrue(result.containsKey(TABLE_RELATION_RETRY_COUNT));
+
+		String errorMessage = (String) result.get(TABLE_RELATION_EXCEPTION_OUTPUT);
+		assertTrue(errorMessage.startsWith("RETRYABLE:"));
+		assertEquals(1, result.get(TABLE_RELATION_RETRY_COUNT));
+	}
+
+	@Test
+	void testDatabaseQueryException_NonRetryable() throws Exception {
+		// Simulate non-retryable database exception (e.g., H2 Boolean conversion issue)
+		DataAccessException nonRetryableException = new DataAccessException(
+				"Feature not supported: converting to class boolean") {
+		};
+		when(businessKnowledgeRecallService.getFieldByDataSetId(anyString())).thenThrow(nonRetryableException);
+
+		setupStateWithValidData();
+		setupSchemaServiceMocks();
+
+		// execute test
+		Map<String, Object> result = tableRelationNode.apply(state);
+
+		// Verify error handling
+		assertNotNull(result);
+		assertTrue(result.containsKey(TABLE_RELATION_EXCEPTION_OUTPUT));
+
+		String errorMessage = (String) result.get(TABLE_RELATION_EXCEPTION_OUTPUT);
+		assertTrue(errorMessage.startsWith("NON_RETRYABLE:"));
+		assertEquals(1, result.get(TABLE_RELATION_RETRY_COUNT));
+	}
+
+	@Test
+	void testRetryCountIncrement() throws Exception {
+		setupStateWithValidData();
+		setupSchemaServiceMocks();
+		// Set existing retry count
+		state.updateState(Map.of(TABLE_RELATION_RETRY_COUNT, 2));
+
+		DataAccessException exception = new DataAccessException("Network error") {
+		};
+		when(businessKnowledgeRecallService.getFieldByDataSetId(anyString())).thenThrow(exception);
+
+		// execute test
+		Map<String, Object> result = tableRelationNode.apply(state);
+
+		// Verify retry count increment
+		assertEquals(3, result.get(TABLE_RELATION_RETRY_COUNT));
+	}
+
+	@Test
+    void testSemanticModelServiceException() throws Exception {
+        // Simulate semantic model service exception
+        when(businessKnowledgeRecallService.getFieldByDataSetId(anyString()))
+                .thenReturn(new ArrayList<>());
+        when(semanticModelRecallService.getFieldByDataSetId(anyString()))
+                .thenThrow(new DataAccessException("Semantic model service error") {});
+
+        setupStateWithValidData();
+        setupSchemaServiceMocks();
+
+        // execute test
+        Map<String, Object> result = tableRelationNode.apply(state);
+
+        // Verify error handling
+        assertTrue(result.containsKey(TABLE_RELATION_EXCEPTION_OUTPUT));
+        assertEquals(1, result.get(TABLE_RELATION_RETRY_COUNT));
+    }
+
+	private void setupSuccessfulMocks() throws Exception {
+		// Simulate successful business knowledge recall
+		List<BusinessKnowledgeDTO> businessKnowledges = List.of(createBusinessKnowledge("term1", "description1"));
+		when(businessKnowledgeRecallService.getFieldByDataSetId(anyString())).thenReturn(businessKnowledges);
+
+		// Simulate successful semantic model recall
+		List<SemanticModelDTO> semanticModels = List.of(createSemanticModel("user_age", "用户年龄"));
+		when(semanticModelRecallService.getFieldByDataSetId(anyString())).thenReturn(semanticModels);
+
+		setupSchemaServiceMocks();
+	}
+
+	private void setupSchemaServiceMocks() {
+		// Simulate Schema service
+		doNothing().when(baseSchemaService).extractDatabaseName(any(SchemaDTO.class));
+		doNothing().when(baseSchemaService).buildSchemaFromDocuments(any(), any(), any());
+
+		SchemaDTO mockSchema = new SchemaDTO();
+		when(baseNl2SqlService.fineSelect(any(SchemaDTO.class), anyString(), any(List.class))).thenReturn(mockSchema);
+	}
+
+	private void setupStateWithValidData() {
+		List<Document> tableDocuments = List.of(new Document("table content"));
+		List<List<Document>> columnDocuments = List.of(List.of(new Document("column content")));
+
+		state.updateState(Map.of(INPUT_KEY, "test query", EVIDENCES, List.of("evidence1"),
+				TABLE_DOCUMENTS_FOR_SCHEMA_OUTPUT, tableDocuments, COLUMN_DOCUMENTS_BY_KEYWORDS_OUTPUT, columnDocuments,
+				AGENT_ID, "123", TABLE_RELATION_RETRY_COUNT, 0));
+	}
+
+	private BusinessKnowledgeDTO createBusinessKnowledge(String term, String description) {
+		BusinessKnowledgeDTO dto = new BusinessKnowledgeDTO();
+		dto.setBusinessTerm(term);
+		dto.setDescription(description);
+		return dto;
+	}
+
+	private SemanticModelDTO createSemanticModel(String originalFieldName, String agentFieldName) {
+		// 使用 SemanticModelDTO 的实际构造函数和字段
+		SemanticModelDTO dto = new SemanticModelDTO();
+		dto.setOriginalFieldName(originalFieldName);
+		dto.setAgentFieldName(agentFieldName);
+		dto.setFieldSynonyms("年龄,岁数");
+		dto.setFieldDescription("用户的年龄信息");
+		dto.setFieldType("INTEGER");
+		dto.setOriginalDescription("数据库中的用户年龄字段");
+		dto.setDefaultRecall(true);
+		dto.setEnabled(true);
+		return dto;
+	}
+
+}

